iroha/src/kura.rs

```diff
@@ -180,126 +196,248 @@ impl Default for Mode {
 #[derive(Debug)]
 pub struct BlockStore {
     path: PathBuf,
+    blocks_per_file: NonZeroU64,
 }
 
 impl BlockStore {
     /// Initialize block storage at `path`.
     ///
     /// # Errors
     /// - Failed to create directory.
-    pub fn new(path: &Path) -> Result<BlockStore> {
-        if fs::read_dir(path).is_err() {
-            fs::create_dir_all(path).wrap_err("Failed to create Block Store directory.")?;
+    pub async fn new(path: &Path, blocks_per_file: NonZeroU64) -> Result<BlockStore> {
+        if fs::read_dir(path).await.is_err() {
+            fs::create_dir_all(path)
+                .await
+                .wrap_err("Failed to create Block Store directory.")?;
         }
         Ok(BlockStore {
             path: path.to_path_buf(),
+            blocks_per_file,
         })
     }
 
-    fn get_block_filename(block_height: u64) -> String {
-        format!("{}", block_height)
+    /// Returns expected filename of datafile containing block, provided its height
+    async fn get_block_filename(
+        dir: &Path,
+        block_height: NonZeroU64,
+        blocks_per_file: NonZeroU64,
+    ) -> Result<String> {
+        const INITIAL_DATA_FILE: &str = "1";
+        let storage_indices = storage_files_base_indices(dir)
+            .await
+            .wrap_err("Failed to list storage files")?;
+        let max_base = match storage_indices.iter().last() {
+            Some(max_base) => *max_base,
+            None => return Ok(INITIAL_DATA_FILE.to_owned()),
+        };
+        if block_height <= max_base {
+            return Err(eyre::eyre!("An attempt to random-access persistent storage (file except for the last one queued)."));
+        }
+        if (block_height.get() - max_base.get()) < blocks_per_file.get() {
+            return Ok(max_base.to_string());
+        }
+        Ok(block_height.to_string())
```

Comment:
Just a refactoring. It's up to you to adopt
```suggestion
        if let Some(max_base) = storage_indices.iter().last() {
            if *max_base < block_height {
                if block_height.get() - max_base.get() < blocks_per_file.get() {
                    return Ok(max_base.to_string());
                }
                return Ok(block_height.to_string());
            }
            return Err(eyre::eyre!("An attempt to random-access persistent storage (file except for the last one queued)."));
        }
        Ok(INITIAL_DATA_FILE.to_owned())
```

Reply:
Both styles are ok, as long as the code is readable(nesting is not too deep, etc.). We don't have a preference in our style guide, so this should probably stay up to the PR author.

Reply:
Here probably both styles are okay, but on bigger functions it is hard to go through nested code. I agree that we dont have preferences, but style is also a discussable thing.

Reply:
I'll pick one with less nesting - it's easier to read IMO.